Make lazy codegen honor per-operator-headers flag

wconstab opened the pull request:
Summary:
- per-operator-headers is a strict build mode where compulation units aren't allowed
to depend on bulk headers like ATen/Functions.h, but must instead depend only on the
specific operator headers used.  (In other configurations, the reverse is required).

Test Plan: CI to make sure nothing breaks for existing backends, and rebased next diff manual test to make sure it actually helps

Differential Revision: D35002666

bdhirsh commented on .jenkins/pytorch/test.sh:
```diff
@@ -274,6 +274,14 @@ test_libtorch() {
       "$TORCH_BIN_DIR"/test_jit  --gtest_filter='-*CUDA' --gtest_output=xml:$TEST_REPORTS_DIR/test_jit.xml
     fi
+
+    # Run Lazy Tensor cpp tests
```
mostly just wondering - were the cpp tests no running in CI before this? And any reason to only enable them for cuda builds on CI? (once the XLA plugin is ready, I'm guessing we want it to run on XLA too?)

wconstab replied:
correct, it was an oversight, but not a big deal since we mostly looked at the lazy_tensor_staging CI, and also this test was running in sandcastle.  But it did confuse me for a while when I was trying to find the output of the expanded test suite in my later diff!

also note, this change is supposed to be landed first in https://github.com/pytorch/pytorch/pull/74449

I'm not sure why the rebase didn't work as intended, but this diff should only be adding the per-operator-headers stuff, and i'll land the other one first and make sure this one is clean.

bdhirsh commented on caffe2/CMakeLists.txt:
```diff
@@ -410,6 +415,7 @@ if(NOT INTERN_BUILD_MOBILE OR NOT BUILD_CAFFE2_MOBILE)
       $<$<BOOL:${SELECTED_OP_LIST}>:--selected-op-list-path="${SELECTED_OP_LIST}">
       --force_schema_registration
       --gen_lazy_ts_backend
+      ${GEN_PER_OPERATOR_FLAG}
```
hmm I might just be misunderstanding. But why does the actual codegen need to know if we're building the per-operator flag or not?

I think the way it works today is that the codegen will always emit the per-operator headers. The codegen templates files themselves include an IFDEF like this:
```
#ifndef AT_PER_OPERATOR_HEADERS
#include <ATen/Operators.h>
#else
${operator_headers}
#endif
```

Example template: https://github.com/pytorch/pytorch/blob/14bf20cd922c3ba33c32343c19fd9ac490d4f7a6/aten/src/ATen/templates/RegisterFunctionalization.cpp#L9

wconstab replied:
I think I didn't realize this logic existed, but it is still not enough since I also had to fix this (see below)

```'ops_headers': '#include <ATen/Functions.h>' if not per_operator_headers else '',```



Perhaps we could move ATen/Functions.h into the template ifdef and then things would have been OK.

wconstab replied:
I think I didn't realize this logic existed, but it is still not enough since I also had to fix this (see below)

```'ops_headers': '#include <ATen/Functions.h>' if not per_operator_headers else '',```



Perhaps we could move ATen/Functions.h into the template ifdef and then things would have been OK.

wconstab replied:
> I think the way it works today is that the codegen will always emit the per-operator headers. The codegen templates files themselves include an IFDEF like this:

I think this isn't quite correct, the `RegisterFunctionalization` template is a nice thing to emulate but it is not how `RegisterDispatchKey` works currently.  RegisterDispatchKey currently has its `per_operator_headers` thing hardcoded to False in gen_backend_stubs.py (probably since that is presumed only used out of tree), and RegisterDispatchKey.cpp doesn't have the nice #ifndef that you pointed out in RegisterFunctionalization.

I was starting to mess with this, and I am kind of wondering if it's ok to land as-is and then update how it works.  I mainly don't want to add any more delay to landing the TS backend since that is blocking other stuff downstream.

bdhirsh replied:
ok yep, I definitely don't want to hold up the PR on this if landing this PR is actively blocking other things.

wconstab committed:
Make lazy codegen honor per-operator-headers flag (#74450)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/74450

- per-operator-headers is a strict build mode where compulation units aren't allowed
to depend on bulk headers like ATen/Functions.h, but must instead depend only on the
specific operator headers used.  (In other configurations, the reverse is required).

Test Plan: CI to make sure nothing breaks for existing backends, and rebased next diff manual test to make sure it actually helps

Reviewed By: ezyang

Differential Revision: D35002666

fbshipit-source-id: 37d1142e165a1f326df2f463c261c5058c244d10
